### PR TITLE
Autotune enabled in GQC.

### PR DESCRIPTION
**Pull request by batinkov**
A new UI component with the corresponding model are added. They enables Autotune in GQC. The autotuning can be performed only when the vehicle is flying. The new Autotune UI and the old interface are put under a switch and only one of them is available to the pilot at at time.

https://user-images.githubusercontent.com/1894862/138436091-178567c3-e7c3-4fd4-a1c0-719a3105680a.mp4

### CODE REVIEW

File: src/Vehicle/Autotune.cpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+
+//-----------------------------------------------------------------------------
+void Autotune::handleAckFailure()
+{
+    stopTimers();
+    _autotuneInProgress = false;
+    _disarmMessageDisplayed = false;
+    _autotuneStatus = tr("Autotune: Failed");
```
[the remark this answers is not in the document]
Comment (hamishwillee) [reply src/Vehicle/Autotune.cpp:134]: @bkueng @DonLakeFlyer Can a frame know it supports the feature in advance? If so we could have a protocol bit, or component information.

**Comment by hamishwillee** [comment]
@batinkov @bresch Did you guys already have a plan for the "autotuning guide" (referred to in the autotuning prompt)?

Here are some questions/things that spring to mind

1. What vehicles can be auto-tuned?

   - Sounds like you can tune fixed wing or multicopter, or VTOL in either mode. Correct?
   - Can you tune VTOL transition, boat, submarines, blimps? My "guess" is that you can tune the rate and attitude controllers on anything that has them (correct)?, but I "guess" that maybe you don't have these on a boat say?

2. Does this just work for PX4 or does the same UI work for ArduPilot?
3. What happens if the autotuning isn't supported?
   - It sounds like autotuning will be attempted and it fails for any reason you'll get a notification.
   - What form does that notification take?
4. Can this be tested in Gazebo? 
   - I tried and the UI isn't showing up in QGC daily today but wanted to be sure.
5. Tuning has to be done in the air. Is there any particular height recommendation? i.e. how much variation in height might you expect during tuning?
6. It looks like you have to land and disarm to apply the autotuning parameters - is that correct?
7. How long does it normally take?
   - Looks like about 30 seconds in the video
   - Does it take longer for different airframes?
8. How violent is it - i.e. are there frames that are likely to take damage from it?
9. Would I be correct that the MAVLink protocol is to simply send [MAV_CMD_DO_AUTOTUNE_ENABLE](https://mavlink.io/en/messages/common.html#MAV_CMD_DO_AUTOTUNE_ENABLE) to start enabling? Over this you have layered progress notification using STATUSTEXT? 
   - YOu could perhaps also have done this with https://mavlink.io/en/services/command.html#long_running_commands
10. What does it mean to say that autotuning might fail for a particular frame - i.e. is this because the

@bresch There are three tuning guides here: http://docs.px4.io/master/en/config_fw/ and a couple here http://docs.px4.io/master/en/config_mc/

Thoughts on how best to update these? It might be worth starting the relevant ones with a section explaining autotuning and pointing to the bits you still have to do manually.

File: src/Vehicle/Autotune.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (bkueng) [reply src/Vehicle/Autotune.cpp:134]: I was thinking to add it to a new comp info file, but your other options would work here too.
There's more small bits like this, for example if sensor calibration reset to factory is supported. Comp info would make this simple to extend and we can add as many as we want.

**Comment by bresch** [comment]
Before starting, let me mention that I'm planning to port the following guide to the PX4 doc:
https://docs.auterion.com/skynode/advanced-configuration/controllers-auto-tuning

> What vehicles can be auto-tuned?
Sounds like you can tune fixed wing or multicopter, or VTOL in either mode. Correct?

Correct

> Can you tune VTOL transition, boat, submarines, blimps? My "guess" is that you can tune the rate and attitude controllers on anything that has them (correct)?, but I "guess" that maybe you don't have these on a boat say?

VTOL transition: no, just tune MC and FW modes independently
boat, submarines, blimps: as you said, if they have a rate controller, we should be able to enable it, but currently, it's only interfaced for FW and MC, so no, it won't works atm.

> Does this just work for PX4 or does the same UI work for ArduPilot?

I never tried

> What happens if the autotuning isn't supported?
It sounds like autotuning will be attempted and it fails for any reason you'll get a notification.
What form does that notification take?

Yes, it should return "Autotuning: ack error"

> Can this be tested in Gazebo?

Yes

> I tried and the UI isn't showing up in QGC daily today but wanted to be sure.

That's strange, I would need to check

> Tuning has to be done in the air. Is there any particular height recommendation? i.e. how much variation in height might you expect during tuning?

The drone can fly in alt/pos/auto modes during the auto-tuning, so the altitude shouldn't vary much. Also, if in pos/auto mode, the horizontal displacement should be small. In alt mode, it'll drift with wind and manual correction attempts will abort auto-tuning at the moment, this is still something we can improve if needed.

> It looks like you have to land and disarm to apply the autotuning parameters - is that correct?

You can define that behavior in the `[MC|FW]_AT_APPLY` parameters. You can then choose to apply them in air or after disarm. For MC, apply after disarm is generally safer since you can then takeoff carefully (default behavior for MC). For FW it's better to apply in air because you'll get some time to recover if the tuning isn't great (default behavior for FW); furthermore, there is a logic to restore the previous parameters if the energy of the actuator signal is too large with the new gains.

> How long does it normally take?
Looks like about 30 seconds in the video

The time is 5s-20s (aborted if did not converge in 20s) per axis + 2s pause between each axis + 4s of testing if the new gains are applied in air. So, minimum of 21s (= 3x5+3x2) without in-air test, 25s with in-air test and a worst case of 70s (if each axis takes exactly 20s to converge). Given the flight we've made so far, it usually takes no more than 40s.

> Does it take longer for different airframes?
In FW mode, the axis to be tuned can be selected, so a user could make a run to just tune the roll axis for example. In that case, it will take less time. Other than that, it cannot take more than the worst case 70s.

> How violent is it - i.e. are there frames that are likely to take damage from it?

It isn't violent on a standard MC but some larger VTOL planes in MC mode might not like large kicks, this is why the amplitude of the injected control signal (required for system identification) can be adjusted via a parameter ([MC_AT_SYSID_AMP](https://docs.px4.io/master/en/advanced_config/parameter_reference.html#MC_AT_SYSID_AMP) and `FW_AT_SYSID_AMP`) if needed.

> Would I be correct that the MAVLink protocol is to simply send MAV_CMD_DO_AUTOTUNE_ENABLE to start enabling? 

Yes

> Over this you have layered progress notification using STATUSTEXT?

AFAIK, the command is sent periodically and the status is returned in the progress field of [COMMAND_ACK](https://mavlink.io/en/messages/common.html#COMMAND_ACK)

> You could perhaps also have done this with https://mavlink.io/en/services/command.html#long_running_commands
What does it mean to say that autotuning might fail for a particular frame

The mathematical model used to estimate the dynamics of the drone assumes this it is a linear system with no coupling between the axes (SISO) and with a limited complexity (2 poles and 2 zeros). If the real drone is too far from those conditions, the model will not be able to represent the real dynamics of the drone.

**Comment by hamishwillee** [comment]
Thanks @bresch 

Just my two bits, but I find it a little odd that the autotune button is hidden inside the rate/attitude controller sections. I understand why, but were it me I would 
1. Enable it by default on frames where we think it is supported
2. Remove it altogether on flight stacks and frames where we know it is not supported
3. Maybe turn off the airmode option while autotune is enabled
4. Have it at the top outside above the tabs as the first thing people see - possibly with slight greying of the tabs that it affects when it is turned on - indicating that the other tabs still need to be looked at.

Why? Well this is what most people will want to do and right now it isn't obvious that the autotune buttons on the tabs are linked. Maybe get a UX designer to comment.

> AFAIK, the command is sent periodically and the status is returned in the progress field of COMMAND_ACK

That's one way of implementing a client side the long running protocol (LRP). Essentially the LRP says you send a command, it acks periodically with progress until it completes, then it acks with acknowledged. If you ping it with the same message while in progress you get the progress ACK resent - which is what happens here. But you're ignoring the final ACK and just using the last received progress being 100% or disarmed to indicate completion. 

This almost certainly won't work on ArduPilot because I don't believe they use the progress notifications at all. 

Can you try out the gazebo? I tried fresh PX4 master and QGC daily today. The autotune button did not appear for multicopter. For plane the PID section did not appear.

**Comment by hamishwillee** [comment]
**This won't work on ArduPilot, and that is a problem**. They interpret this command as "enable autotune". That forces the fixed wing vehicle into an autotune mode. 

While I prefer the PX4 approach, doing something different than the intent is non-standard. It means that a GCS has to do something different for every single platform.

Upshot, if we're going to do this, we probably need to agree a new command and document the precise semantics. FYI @julianoes @DonLakeFlyer

**Comment by bresch** [comment]
> Can you try out the gazebo? I tried fresh PX4 master and QGC daily today. The autotune button did not appear for multicopter. 

It tried QGC daily in jMavSim and I also don't get the autotuning UI

**Comment by batinkov** [comment]
@bresch I don't know why you don't get the Autotune interface. I get it and it works as expected with PX4 (proprietary) and PX4 (open source) and gazebo_typhoon_h480, px4_sitl gazebo_standard_vtol, jmavsim, and px4_sitl gazebo. All possible combinations give me the interface and Autotune always works. Can someone else confirm they don't get the Autotune interface?

**Comment by batinkov** [comment]
@DonLakeFlyer I saw your comments up and I haven't forgotten about them. It's pretty crazy lately at work and that's why I'm pushing back those fixes.

**Comment by batinkov** [comment]
Finally I'm not the right person to judge about MAVLink but it looks @hamishwillee is right about the Autotune implementation in PX4. Perhaps we need to change the polling mechanism in PX4 implementation and then QGC needs to change accordingly.

**Comment by hamishwillee** [comment]
Me, I don't get the autotune interface. Are you using ubuntu and current gazebo?

**Comment by batinkov** [comment]
@hamishwillee I'm using Ubuntu 18.04 LTS and gazebo 9.19.0.

**Comment by hamishwillee** [comment]
@batinkov How? QGC doesn't work for me on Ubuntu 18.04.
 From https://docs.qgroundcontrol.com/master/en/getting_started/download_and_install.html#ubuntu

image

**Comment by hamishwillee** [comment]
@batinkov @bresch I did a completely clean install of PX4 and QGC daily on Ubuntu 20.04 and I am now seeing the autotune button!

**Comment by DonLakeFlyer** [comment]
Ramon fixed the 20.04 requirement a few days back. With that the docs are out of data and 18.04 is fine now.

**Comment by mrpollo** [comment]
@hamishwillee @DonLakeFlyer https://github.com/mavlink/qgc-user-guide/pull/306

**Pull request merged**